dilated-file-transfer.md

```diff
@@ -0,0 +1,484 @@
+    XXX: that is:
+    XXX: a "receive-only" client can simply never send an offer
+    XXX: a "send-only" client simply (automatically) rejects any offer
+
+    XXX: "being explicit" has the advantage that a "receive-only" peer that connects to another "receive-only" peer will discover that quickly, and can (properly) fail -- otherwise, they'll only notice when one side gets bored and quits
+
+    XXX: two "send-only" peers that connect will fail fairly quickly -- each side should disconnect with a protocol-error when their receive the first offer
+
+    XXX: a "connect" peer that contacts either a "receive-only" or "send-only" can benefit from "being explicit" by disabling parts of the UI (for example) that won't work (e.g. contacting a "send-only" peer means the user can't browse/drop files)
+
+The `"features"` key points at a list of message-formats / features understood by the peer.
+This allows for existing messages to be extended, or for new message types to be added.
+Peers MUST _accept_ messages for any features they declare in `"features"`.
+Peers MUST only send messages / attributes for features in the other side's list.
+Only one format exists currently: `"core"`.
+
+   XXX:: maybe just lean on "version" for now? e.g. version `2` could introduce "features"?
+
+See "Example of Protocol Expansion" below for discussion about adding new attributes (including when we might increment the `"version"` instead of adding a new `"feature"`).
+
+The `"permission"` key specifies how to proceed with offers.
+Using mode `"ask"` tells the sender to pause after transmitting the first metadata message and await an answer from the peer before streaming further data.
+Using mode `"yes"` means the peer will accept all incoming transfers so the sender should not pause after the metadata (and instead immediately start sending data messages).
+Although a peer could implement "yes" mode by simply sending an accept message for each offer without user interaction, setting this mode cuts down latency for "one-way" transfers (see [Discussion](#discussion))
+
+
+## Protocol Details
+
+See the Dilation document for details on the Dilation setup procedure.
+Once a Dilation-supporting connection is open, we will have a "control" subchannel (subchannel #0).
+Either peer can also open additional subchannels.
+
+All control-channel messages are encoded using `msgpack`.
+   --> XXX: see "message encoding" discussion
+
+Control-channel message formats are described using Python pseudo-code to illustrate the data types involved.
+They are actually an encoded `Map` with `String` keys (to use `msgpack` lingo) and values as per the pseudo-code.
+
+All control-channel messages contain a integer "kind" field describing the sort of message it is.
+(That is, `"kind": 1` for example, not the single-byte tag used for subchannel messages)
```

Comment:
I am strongly in favor of using strings instead of integers for all enums and other type discriminants, here and below.

Reply:
Why?

Reply:
Because integers are just another layer of indirection. You need to specify which integer means what, instead of giving things a declarative name. Many serialization frameworks (e.g. serde, gson) support automatic naming based on the identifier name in the code, so this reduces both programming effort and possibilities for mistakes. Furthermore, the space of strings is a lot larger than just single byte integers, so we don't need to do things like having to reserve IDs for purposes and stuff like that.

The only advantage of using integers I can see is size/efficiency, and I don't think this is a concern for us here.

Reply:
Using a fixed-length field for this in subchannel messages does have advantages since the vast majority of data down those channels will simply be payload bytes of file data. So, you save an additional pass over that data and a second copy in memory.

No matter what, the spec still needs to define a mapping from "thing" to "meaning" whether "thing" is `0x02` or `"payload-data"`.

Ideally we'd just get rid of this part _entirely_ and base the meaning of records on their position (e.g. "first one is the offer", etc) which would allow writing the file-data blobs directly. The implications of this if a stream got 'out of sync' -- e.g. one side thinks "this is msgpack data" while other thinks "it's file data" could be interesting to consider -- but also, that shouldn't happen (unless e.g. metadata is wrong about the file size etc).